### PR TITLE
fix: harden proxy compatibility

### PR DESCRIPTION
**Pull request by sozercan**
## Summary
- Harden proxy compatibility and reliability.
- Add regression coverage for important edge cases.
- Refresh supporting documentation.

## Validation
- `make test`
- `make build`
- `make vet`
- `make lint`
- `golangci-lint run --new-from-rev=de0bb7a4bcce7e66e33d23efc024dc22a39dfd97`
- `go test ./proxy -race -count=1`

### CODE REVIEW

File: proxy/streaming.go [hunk 1]
```diff
@@ -118,7 +124,7 @@ func StreamOpenAIPassthroughWithFinalResponse(
 	}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply proxy/streaming.go:129]: Both issues are fixed in `StreamOpenAIPassthroughWithFinalResponse` (proxy/streaming.go). In the `processData` closure, `onUsage` is now actually invoked with `chunk.Usage` when both are non-nil — it was previously wired up but never called — so streaming usage is recorded; the callback is non-nil on the normal path because `withRequestLog` always attaches a `RequestSummary` to the context. The nil-pointer crash is fixed by guarding `aggregator.addChunk` with an `aggregator != nil` check. The aggregator is allocated only when `onFinalResponse != nil`, which is exactly the default-config case where it was nil while `onUsage` was set, so the first SSE chunk dereferenced a nil aggregator.

Added regression test `TestStreamingChatCompletionsPassthroughThroughServer` (server/server_test.go) that drives `stream:true` through `server.New` with default config (onFinalResponse nil, onUsage non-nil) and calls the handler via `ServeHTTP` on an httptest recorder. There is no recover in the proxy/server request path, so the old code panicked here — I confirmed this by reverting just the nil-guard: the test fails with a nil-pointer dereference at `openAIResponseAggregator.addChunk` (streaming.go:592), and passes once restored. It asserts verbatim forwarding through `data: [DONE]` and that streamed `prompt_tokens`/`completion_tokens`/`total_tokens` reach the request log, which only happens if `onUsage` actually fires. Passes under `-race`.

Fixed in 38d806e.

File: proxy/streaming.go [hunk 2]
```diff
@@ -93,17 +95,21 @@ func StreamOpenAIPassthroughWithFinalResponse(
 		flusher = f
 	}
 
-	if onFinalResponse == nil {
+	onUsage := firstOpenAIUsageCallback(onUsageCallbacks)
+	if onFinalResponse == nil && onUsage == nil {
```
[the remark this answers is not in the document]
Comment (sozercan) [reply proxy/streaming.go:99]: Fixed. `StreamOpenAIPassthroughWithFinalResponse` now guards both callbacks independently in `processData`: `onUsage` is only called when non-nil, and `aggregator.addChunk` only when the aggregator was actually allocated. The aggregator is still allocated solely when `onFinalResponse != nil`, and the entry gate already takes the parsing path whenever `onUsage` is non-nil. So the default-config path — `onUsage` non-nil via the `RequestSummary` that `withRequestLog` attaches, `onFinalResponse` nil because tool optimizers are off (`openAIChatStreamFinalResponseCallback` returns nil) — now streams through instead of dereferencing a nil aggregator on the first data chunk, which previously aborted every streamed `/v1/chat/completions` request.

Added `TestStreamingChatCompletionsPassthroughThroughServer` (server/server_test.go), which builds the proxy via `server.New` and drives a real streamed `/v1/chat/completions` through the full handler chain with optimizers off. It asserts the body is forwarded verbatim through `data: [DONE]` and that streaming usage lands in the request log. I verified it reproduces the original panic: removing the nil guard makes the test fail with a nil-pointer dereference at `openAIResponseAggregator.addChunk` (streaming.go:592) from the processData closure with `onFinalResponse=0x0`. Passes under `-race` with the fix in place.

Fixed in 38d806e. Thanks for the clear P1 writeup — the default-config reachability was exactly right.

File: proxy/streaming.go [hunk 3]
```diff
@@ -239,14 +255,33 @@ func StreamOpenAIToAnthropicWithFinalResponse(
...
+		}
 		if aggregator != nil {
-			aggregator.addChunk(chunk)
+			if aggregator != nil {
+				aggregator.addChunk(chunk)
+			}
+			if onUsage != nil && chunk.Usage != nil {
+				onUsage(chunk.Usage)
+			}
 		}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply proxy/streaming.go:271]: Fixed. The callback in `StreamOpenAIToAnthropicWithFinalResponse` now calls `onUsage` exactly once (guarded by `onUsage != nil && chunk.Usage != nil`) and `aggregator.addChunk` exactly once (guarded by `aggregator != nil`). The duplicate `onUsage` call and the redundant nested `if aggregator != nil` that were inside the aggregator block are both removed, so usage is no longer double-counted when an aggregator is present. This now matches the clean form already used in the Gemini translator (`proxy/gemini_streaming.go`). Build is clean and the proxy stream/usage tests pass.

Fixed in 38d806e.

File: proxy/responses_websocket.go [hunk 4]
```diff
@@ -186,13 +208,151 @@ func (h *ProxyHandler) HandleResponsesWebSocket(w http.ResponseWriter, r *http.R
...
+}
+
+func (s *responsesWebSocketSession) clearInflightCancel(context.CancelFunc) {
+	if s == nil {
+		return
+	}
+	s.inflightMu.Lock()
+	defer s.inflightMu.Unlock()
+	s.inflightCancel = nil
+}
```
[the remark this answers is not in the document]
Comment (sozercan) [reply proxy/responses_websocket.go]: Good catch on the intent — `clearInflightCancel` now compare-and-clears so it can't nil out a newer cancel. One Go detail worth calling out: `context.CancelFunc` is a func value, and func values aren't comparable with `==`, so the literal suggestion (compare the stored cancel against the passed-in one) would be a compile error. To honor the same compare-and-clear semantics I used a generation token: `setInflightCancel` bumps `s.inflightGen` and returns it, and `clearInflightCancel(gen)` only nils `inflightCancel` when `s.inflightGen == gen`. Both call sites pass the captured `gen` to the deferred clear, so the `handleCreateRequest` → `maybeAutoCompactHistory` nesting unwinds LIFO: the inner clear nils its own generation, and the outer clear becomes a no-op instead of nilling the newer cancel. `cancelInflight` (ping loop and shutdown) is unchanged and only reads.

On severity, to be straight about it: the websocket read loop processes frames serially (no goroutine-per-frame), `inflightCancel` has a single writer goroutine, and `cancelInflight` only reads — so the literal stale-clear race isn't reachable today. The new clear also nils in a strict subset of the old unconditional clear's cases, so it can only retain a cancel longer, never drop one earlier; cancellation behavior is unchanged or strictly safer. This is latent/defensive hardening that makes the nested-unwind identity semantics explicit and correct, not the repair of an observable live race. No deadlock: `cancelInflight` releases `inflightMu` before invoking `cancel()`, and set/clear never call out under the lock. build and vet are clean and the WebSocket suite passes under `-race`.

Fixed in 38d806e.

File: proxy/responses_handler.go [hunk 5]
```diff
@@ -2041,20 +2133,66 @@ func (h *ProxyHandler) postResponsesCompactWithFallback(ctx context.Context, bod
 // applyResolvedCompactModel rewrites bodyBytes' "model" field to the budget's
 // resolved fallback model when one has been recorded. Failures fall back to
 // the original body so a malformed request still gets the prior fallback path.
+func sanitizeResponsesModelFallbackBody(body []byte) []byte {
+	var payload map[string]json.RawMessage
+	if err := json.Unmarshal(body, &payload); err != nil {
+		return body
+	}
+	delete(payload, "previous_response_id")
```
[the remark this answers is not in the document]
Comment (sozercan) [reply proxy/responses_handler.go:2141]: I dug into this one and I think the current behavior is the least-broken choice on this specific path, so I'm going to push back — but it's a fair question and there's a narrow real caveat, so let me lay out the reasoning.

This sanitizer only runs on the **unsupported-model fallback** path: it triggers only when the upstream rejects the *originally requested* model as unsupported (`isUnsupportedResponsesModelError`, responses_handler.go:2079), and then retries against a **different** model chosen by `pickResponsesCompatibleModel`, which explicitly excludes the requested model (responses_handler.go:2085, and the `model.publicID == exclude` skip at :2999). So by construction the retry targets a different model/deployment than the one that minted `previous_response_id`.

On the `/responses` API a `previous_response_id` is a handle to server-side state bound to the deployment that created it; it isn't portable across deployments (this matters most for Azure, a primary target). So if we *kept* the ID on the fallback retry, the fallback model would almost certainly reject it ("previous response not found"), turning the recovery retry — the entire point of the fallback — into a guaranteed hard error. Dropping it degrades gracefully to a valid (context-thinner) response instead. Note it's stripped alongside `reasoning` items and `encrypted_content` (:2152, :2158), which are the same class of model-bound opaque handles; keeping the ID while dropping those would be internally inconsistent.

Also worth noting: the sanitizer preserves `payload["input"]` (:2143-2170) — it filters reasoning/encrypted_content but not the conversation messages. For the replay-style clients these proxy-managed paths target, the full history travels in `input`, so `previous_response_id` is redundant here.

The one honest caveat: if a prior ID was itself produced by the fallback model on an earlier fallback turn, it *would* be valid on that model, and a thin non-replaying client would lose that turn's server-side state. But the proxy keeps no response-ID→model provenance, so it can't distinguish that case from the cross-model-invalid one — and defaulting to "keep" would break the common case (cross-model retry) to rescue the rare one.

Happy to revisit if you're seeing a concrete client/deployment where the fallback model accepts the original model's `previous_response_id` — that would change the calculus. Leaving the behavior as-is for now.
